plugins/home/src/components/CustomHomepage/types.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ReactNode } from 'react';
+import { RJSFSchema } from '@rjsf/utils';
+import { Layout } from 'react-grid-layout';
+import { z } from 'zod';
+
+const RSJFTypeSchema: z.ZodType<RJSFSchema> = z.any();
+const ReactNodeSchema: z.ZodType<ReactNode> = z.any();
+const LayoutSchema: z.ZodType<Layout> = z.any();
+
+/**
+ * @public
+ */
+export const WidgetSettingsSchema = z.object({
+  width: z.number().positive('width must be positive number').optional(),
+  height: z.number().positive('height must be positive number').optional(),
+  minWidth: z.number().positive('minWidth must be positive number').optional(),
+  maxWidth: z.number().positive('maxWidth must be positive number').optional(),
+  minHeight: z
+    .number()
+    .positive('minHeight must be positive number')
+    .optional(),
+  maxHeight: z
+    .number()
+    .positive('maxHeight must be positive number')
+    .optional(),
+  settingsSchema: RSJFTypeSchema.optional(),
+});
+
+/**
+ * @public
+ */
+export type WidgetSettings = z.infer<typeof WidgetSettingsSchema>;
+
+/**
+ * @public
+ */
+export const LayoutConfigurationSchema = z.object({
+  component: ReactNodeSchema,
+  x: z.number().positive('x must be positive number'),
+  y: z.number().positive('y must be positive number'),
+  width: z.number().positive('width must be positive number'),
+  height: z.number().positive('height must be positive number'),
+});
+
+/**
+ * Layout configuration that can be passed to the custom home page.
+ *
+ * @public
+ */
+export type LayoutConfiguration = z.infer<typeof LayoutConfigurationSchema>;
+
+export const WidgetSchema = z.intersection(
+  WidgetSettingsSchema,
+  z.object({
+    name: z.string(),
+    title: z.string().optional(),
+    component: z.intersection(
+      ReactNodeSchema,
+      z.object({
+        type: z.any(),
+        props: z.any(),
+      }),
+    ),
+  }),
+);
+
+export type Widget = z.infer<typeof WidgetSchema>;
+
+const WidgetStoreItemSchema = z.object({
+  id: z.string(),
+  layout: LayoutSchema,
+  settings: z.record(z.string(), z.any()),
+});
+
+export type WidgetStoreItem = z.infer<typeof WidgetStoreItemSchema>;
```

Comment:
Just currious: What is the difference between a `Widget` & a `WidgetStoreItem` & why do you want to differentiate between them?

Reply:
Widget presents component that can be added to the page. StoreItem presents instance of those components that have been added to the grid. Maybe some renaming could help with readability?